pyasic/web/antminer.py

```diff
@@ -59,9 +60,27 @@ async def send_command(
         url = f"http://{self.ip}:{self.port}/cgi-bin/{command}.cgi"
         auth = httpx.DigestAuth(self.username, self.pwd)
         try:
-            async with httpx.AsyncClient(
-                transport=settings.transport(),
-            ) as client:
+            async with httpx.AsyncClient(transport=settings.transport()) as client:
+                if command == "upgrade":
+                    file = parameters.get("file")
+                    keep_settings = parameters.get("keep_settings", True)
+                    if file:
+                        upload_url = f"http://{self.ip}:{self.port}/cgi-bin/firmware_upload.cgi"
+                        async with aiofiles.open(file, "rb") as firmware:
+                            file_content = await firmware.read()
+                            files = {"file": (file.name, file_content, "application/octet-stream")}
+                            upload_response = await client.post(
+                                upload_url,
+                                auth=auth,
+                                files=files,
+                                timeout=settings.get("firmware_upload_timeout", 300)
+                            )
+                        if upload_response.status_code != 200:
+                            return {"success": False, "message": "Failed to upload firmware file"}
+
+                        parameters["filename"] = file.name
+                        parameters["keep_settings"] = keep_settings
```

Comment:
Dont need most of this this.  You should be able to accomplish most of this functionality 1 layer above, in the `web.upgrade_firmware` function.  Read the file there, parse it, then add a file parameter.  Set `filename` and `keep_settings` as keyword args.

Reply:
I have refactored the code to move the file handling